### ISSUE TITLE
Move existing tunnel analyzers to packet analysis framework

### ISSUE THREAD

**Issue by timwoj** [zeek/zeek#1571]
This is a follow-on to #1114. Once the UDP analyzer has been moved, the existing tunnel analyzers (Ayiya, Geneve, Teredo, etc) can be moved over to be packet analyzers. This avoids a trip out of packet analysis, into session analysis, and back into packet analysis. It will also make fixing #1125 much easier.

The analyzers to be moved and their current state on the devel branch:
- [x] Ayiya
- [x] Geneve
- [ ] GTPv1
- [x] Teredo
- [x] VXLAN

**Comment by timwoj** [comment]
One question I'm not entirely sure how to answer is how to deal with AYIYA and Teredo's DPD signatures. AYIYA is generally on a standard port, but the DPD signature makes me believe that it's not necessarily always on that. We don't do DPD at all in the packet analysis world.

**Comment by timwoj** [comment]
Another issue here is that the existing tunnel analyzers have access to the `Connection` object after it's been parsed by the UDP analyzer. This isn't the case if they're packet analyzers, since the object isn't passed along through the analyzer chain in any way.

**Comment by rsmmr** [comment]
> One question I'm not entirely sure how to answer is how to deal with AYIYA and Teredo's DPD signatures. AYIYA is generally on a standard port, but the DPD signature makes me believe that it's not necessarily always on that. We don't do DPD at all in the packet analysis world.

Looking around a bit, it seems that one indeed cannot rely on standard ports (see, e.g., https://networkengineering.stackexchange.com/a/34309). 

I don't think we'd need fully DPD here but I'm wondering if we could give packet analyzers a hook for "claiming" a packet: any packet analyzers that provide the hook would get a chance to look at a packet and say "hey, that's me!". That would work either globally really for *any* packet or, better, we'd have a way to have an analyzer look at a subset of packets (here, e.g.: "all UDP packets"). I don't think it would be too bad performance-wise as any check is probably going to be quick and the data in the cache already anyways.

This could actually be more generally useful for anything that's not easily expressed as next-protocol style number. @0xxon and I were talking about Ethernet yesterday where a specific multicast address may be the signal determine what protocol is spoken.

**Comment by rsmmr** [comment]
> Another issue here is that the existing tunnel analyzers have access to the `Connection` object after it's been parsed by the UDP analyzer. This isn't the case if they're packet analyzers, since the object isn't passed along through the analyzer chain in any way.

Are our analyzers using that for anything? Generally I'd say it'd make sense to remove that because it doesn't really fit conceptually anyways (ignoring backwareds-compatibilty here).

**Comment by grigorescu** [comment]
> That would work either globally for _any_ packet or, better, we'd have a way to have an analyzer look at a subset of packets (here, e.g.: "all UDP packets").

Pie in the sky, I've long wanted a way to be able to do this with BPF. A few times, I've had a use case such as "I'm interested in SYN packets that look like this," but handling the event for all SYN packets is too big of a performance hit.

**Comment by rsmmr** [comment]
Tim and I were chatting about how regexes aren't really a good fit for picking out packets at this layer; BPF on the other hand would be, good point. Still, I think I'd vote for just a callback into the analyzer that can run atribitry C++ code to say which packets it's interested in.

**Comment by kramse** [comment]
> > Another issue here is that the existing tunnel analyzers have access to the `Connection` object after it's been parsed by the UDP analyzer. This isn't the case if they're packet analyzers, since the object isn't passed along through the analyzer chain in any way.
> 
> Are our analyzers using that for anything? Generally I'd say it'd make sense to remove that because it doesn't really fit conceptually anyways (ignoring backwareds-compatibilty here).

VXLAN dont belong to a connection. So I dont think they neeeeed it.

VXLAN ports are usually 4789/UDP - from RFC7384 or 8472/UDP - used by the Linux kernel

I will try to look a bit into this, for VXLAN.

**Comment by rsmmr** [comment]
Not sure where we are here now, adding to milestone for discussion.

**Comment by timwoj** [comment]
I've been working on this for a while now, and have a ton of questions about how tunnels are handled. It's to the point that it might be good to have a wider conversation about how we handle tunnels, and whether we can make some improvements to the whole process while working on this issue.

Sorry for the novel-length comment here, I need to get all of this out of my brain.

Remember, from the original post for this issue, that the problem we're trying to solve here is avoiding a round-trip out of packet-analysis land, into session-analysis, and back into packet-analysis. Tunnel protocols should be able to be handled at the packet analysis level, and since the next protocol in the stack is likely a packet analyzer (ethernet or one of the transport protocols), there's no really good reason to handle them as session protocols. The current code for handling tunnels a couple of assumptions.

The first assumption that the protocol above the tunnel protocol is always going to be UDP (or at least something IP-based). This one probably isn’t a big deal to deal with. For protocols like VXLAN, the RFC pretty much explicitly says that it’s UDP above the tunnel header and so you can always just forward from the UDP analyzer to VXLAN based port number. For something like AYIYA, that isn’t always the case. Thanksfully, the work I’ve already done on this issue adds a new dad-like system for looking at protocol headers. That allows us to look at the first bit of the next header in a generic fashion and find the tunnel protocols. This will probably be necessary for Teredo as well.

The second assumption is that the protocol tree underneath is going to always be something IP-based. Nothing in the RFCs for at least AYIYA, VXLAN, and Geneve (the ones I’ve looked at so far), actually specifies that the packet underneath is any protocol at all. VXLAN’s RFC just shows a “payload”. Geneve shows Ethernet as the next protocol, but nothing for the further protocols. There’s a couple problems here. The first is that the EncapsulationConn object stores an IP/port pair for source and destination. The second is that some of the protocols (such as VXLAN) emit an event when they’re done being processed that includes the IP header for the inner packet.

For the next part, I need to add some context. We already have an IPTunnel analyzer that does some things for us. It helps keep track of tunnel state which means it can track packets in the same tunnel across multiple packets. It creates/maintains the EncapsulationStack for a packet chain. Lastly, it handles forwarding on to the next packet protocol. It’s used as the default analyzer for AYIYA, and works well enough. Ideally, we’d love to handle all of this stuff at a common level, and not have to duplicate it for each tunnel analyzer.

The problem comes from the other analyzers. As mentioned above, some of the analyzers output an event when the entire processing chain is done that includes an IP header for the inner packet. If we’re calling the IPTunnel analyzer from those analyzers, that information gets lost before it gets back to them. So the issue here is how to do forwarding out of the tunnel analyzers. Ideally, `ForwardPacket` would just forward down into the analyzer for the next protocol. For example, Geneve has the protocol information in the header (it’s almost always ethernet), and so we should be able to just forward down to Ethernet. But we still have the problem of information in the inner packet getting lost, plus the problem of whether the inner packet is even IP-based in the first place. Perhaps at this point, EncapsulatingConn could be modified to maintain more information? The outer tunnel analyzer could keep track of where it lies in the encapsulation stack, pull the appropriate EncapsulatingConn out, and use the appropriate information if it exists.

IPTunnel analyzer has to exist in its current form to handle generic ip-in-ip tunnels. Maybe the code can be split up or the other analyzers can inherit from it? They could still implement their own forwarding, but the encapsulation handling and creating internal packets and whatnot would be handled by the parent code. The problem with this is that it continues the first assumption that everything above a tunnel header is IP-based, which it might not be?

TL;DR:
1. Should tunnel analyzers make the assumption that the thing underneath them will always be IP?
2. Can/should EncapsulatingConn store more information about the inner packet, so the tunnel analyzers can use that information when they’re done processing?
3. EncapsulatingConn seems poorly named, as it’s not the encapsulating packet, it’s the encapsulated packet. Also, it’s not really a Connection. It’s... something else.
4. Should IPTunnel analyzer become the parent class for other tunnel analyzers?

**Comment by rsmmr** [comment]
I'm not very familiar with the tunnel code, but let me take a stab at the questions:

> * Should tunnel analyzers make the assumption that the thing underneath them will always be IP?

I think that would be reasonable. First, the current tunnel analyzers are written for IP, so it wouldn't be a new assumption and not breaking anything. Second, I can't even think of an example of a non-IP-based tunnel. I'm sure they exist, but since right now we don't even have non-IP protocols that they would run over in the first place, punting for seems fine to me (plus, hard to design for something that we don't know anyways).

> * Can/should `EncapsulatingConn` store more information about the inner packet, so the tunnel analyzers can use that information when they’re done processing?

That sounds likewise fine me. We shouldn't go overboard with that, but if there's a pretty common piece that multiple protocols could use, I don't see a problem adding that.

> * `EncapsulatingConn` seems poorly named, as it’s not the encapsulating packet, it’s the encapsulated packet. Also, it’s not really a Connection. It’s... something else.

It's not *the* packet either, is it? It's a container with meta information describing a set of encapsulated packets. How about `EncapsulatedFlow`, that's more generic and avoids the `Connection` association. 

> * Should IPTunnel analyzer become the parent class for other tunnel analyzers?

With my 1st answer above, that could work (but I don't know enough about the class to say if semantically it makes sense).

**Comment by timwoj** [comment]
> It's not the packet either, is it? It's a container with meta information describing a set of encapsulated packets. How about EncapsulatedFlow, that's more generic and avoids the Connection association.

Unfortunately it's also used with that name in script land in at least one event, so changing it will be a bit of a mess.

**Comment by timwoj** [comment]
>  think that would be reasonable. First, the current tunnel analyzers are written for IP, so it wouldn't be a new assumption and not breaking anything. Second, I can't even think of an example of a non-IP-based tunnel. 

I think I've got a way around modifying or calling IPTunnel at this point, where the analyzers can just handle their own protocol and forwarding directly. This ends up looking a lot cleaner in the end I think.

One sticking point that keeps coming up for all of the analyzers is calls `ProtocolConfirmation`/`ProtocolViolation`. This appears to be something specific to the session analyzers, and we've never added anything like it for the packet analyzers. I'm not sure if it's even something we want for packet analyzers, but it's used extensively in all of the tunnels, so I'm not sure it can be skipped over.

Some other random design questions:
- We have vectors of ports defined in the existing scripts for each of the tunnel analyzers. Would it be good to move those over to the individual analyzers, or leave them where they are and loop over them to create the forwarding rules?
- Should the DPD-like functionality support configuration of patterns from script land?
- Removing the mapping from UDP to the tunnel analyzers appears to change what ports can be used for BPF filter (in that one of the tests changes output).

**Comment by rsmmr** [comment]
> * We have vectors of ports defined in the existing scripts for each of the tunnel analyzers. Would it be good to move those over to the individual analyzers, or leave them where they are and loop over them to create the forwarding rules?

I'd move them over to the indivudal analyzers, that would be similar to what we do everywhere else.

> * Should the DPD-like functionality support configuration of patterns from script land?

I don't know how a generic mechanism would look like here, so I'd say generally no. But individual analyzers could expose custom options where there's choices in their specific case.

> * Removing the mapping from UDP to the tunnel analyzers appears to change what ports can be used for BPF filter (in that one of the tests changes output).

Not quite sure but the tunnel analyzers might need to register their ports so that they stay in the BPF. Can go into their scripts.

**Comment by rsmmr** [comment]
> One sticking point that keeps coming up for all of the analyzers is calls `ProtocolConfirmation`/`ProtocolViolation`. This appears to be something specific to the session analyzers, and we've never added anything like it for the packet analyzers. I'm not sure if it's even something we want for packet analyzers, but it's used extensively in all of the tunnels, so I'm not sure it can be skipped over.

I don't think we want that generally for packet analyzers (could get expensive and noisy), but don't know what that means for the existing calls ...

**Comment by timwoj** [comment]
> Not quite sure but the tunnel analyzers might need to register their ports so that they stay in the BPF. Can go into their scripts.

This might be a little tricky as the existing port table is indexed by `Analyzer::Tag`. I'll have to think about how to add the tunnel ports into that existing system.

> I don't think we want that generally for packet analyzers (could get expensive and noisy), but don't know what that means for the existing calls ...

The Confirmation/Violation stuff looks like it's almost used like weirds, but with different logging? Do we have any documentation on what it actually does?

**Comment by timwoj** [comment]
I think I caught everything. There might be some amount of nuance I missed though.

### ProtocolConfirmation Event Handling

known_services: if the service has been seen (was added to c$service, probably by dpd below), sends service_info_commit event, which logs the entry to known-services.log. it only does that if it’s a host that’s being tracked though. by default, it only tracks LOCAL_HOSTS.
detect-protocols: ignored because the port is in the list of registered ports. this would be different for the packet analyzer version, since it wouldn’t be registered.
dpd: adds analyzer tag to service list in connection. now done by the new encapsulated_protocol event, but should be extended to do the same thing that this does (with the “-%s” format)
inactivity: sets an inactivity timeout for a connection. only enabled by default for SSH and FTP.

ssl: ignored because it only handles SSL/DTLS
rdp: ignored because it only handles RDP
ssh: ignored because it only handles SSH


### ProtocolViolation Event Handling

dpd (packet-segment-logging): stores packet segment information in the connection record assuming that c$dpd was set previously
dpd (main priority 10): removes analyzer from list of services in c$services and adds a new one with dash for known_services to handle later if needed, stores information in c$dpd
dpd (main priority 5): tracks whether to disable an analyzer because of too many violations
dpd (main priority -5): logs a violation to dpd.log
rdp: writes out the rdp log data, if there is any stored in c$rdp

ssl: ignored because it only handles SSL/DTLS

**Comment by rsmmr** [comment]
So I'd say that for those tunnels running on top of IP (i.e., those for which produce entries in `conn.log` for the outer tunnel connection), it would generally make sense to continue to raise protocol_confirmation/violation events and then have all the normal event handling for those kick in. That said, I know it's tricky to do that because we don't have the session analyzer tags that those events receive. I'm wondering if we could pick the most important cases and add a 2nd path for the information coming out of a packet analyzer, then producing similar results in script-land. I'd say the "important ones" would be management of `c$services`, `known_services` and `detect_protocols`. We could key them off either the existing tunnel events or create a new one, like `tunnel_confirmation` that's the equivalent of `protocol_confirmation`. We could probably just skip the whole `violation` side. The existing confirmation handlers all record strings in the end anyways, so maybe with a little bit of refactoring, we could feed that script code from both session and packet analyzers?

Longer-term I'm wondering if it would make sense to unity all our analyzer tags into a single type so that we have them all inside the same enum namespace. The current separation has been causing trouble and/or replication in a few areas already.

**Comment by timwoj** [comment]
> We could probably just skip the whole violation side.

Another thing I ran into yesterday is that `protocol_violation` is used to disable analyzers that are running into a ton of violations. In the session analysis, this is per-connection obviously since that's how the analyzers are created. I could do the same thing for Teredo pretty easily since it's already storing some per-connection state, but it's not a generic thing.

> Longer-term I'm wondering if it would make sense to unity all our analyzer tags into a single type so that we have them all inside the same enum namespace. The current separation has been causing trouble and/or replication in a few areas already.

I was thinking the same this morning. I almost wonder if it'd be a good idea to shelve the tunnel work and get this resolved first. It's already affecting multiple things in the tunnel work, with protocol_confirmation/violation and the BPF filtering relying on the tags. Fixing it would make the remainder of the work easier.

**Comment by timwoj** [comment]
> Longer-term I'm wondering if it would make sense to unity all our analyzer tags into a single type so that we have them all inside the same enum namespace. The current separation has been causing trouble and/or replication in a few areas already.

I took a stab at this in https://github.com/zeek/zeek/commit/ff8f329c7a87f172758da68011d15f2dea93f097 and tested with the following script:

```
function testing(t : Analyzer::Tag) : bool
{
print t;
print enum_to_int(t);
return T;
}

event zeek_init()
{
testing(PacketAnalyzer::ANALYZER_TEREDO);
testing(Analyzer::ANALYZER_HTTP);
}
```

that outputs

```
PacketAnalyzer::ANALYZER_TEREDO
87
Analyzer::ANALYZER_HTTP
24
```

Passing `-B plugins` shows that all of the analyzers of both kinds get unique tag numbers as well.

**Comment by rsmmr** [comment]
Interesting, that seems easier than I thought! Couple questions:

1. This changes only the C++ side, so we would then still have to make a similar change for scripts, right?
2. This is changing the plugin API. Would it work to keep an alias type inside the packet analyzer namespace to keep it source code compatible? 
3. We have a similar issue with the file analyzers. It's not quite as pressing there (because they don't overlap in semantics the way packet/session does), but would be nice to integrate them as well if that's viable (not sure)?

**Comment by timwoj** [comment]
> We have a similar issue with the file analyzers. It's not quite as pressing there (because they don't overlap in semantics the way packet/session does), but would be nice to integrate them as well if that's viable (not sure)?

Can't all of the tag types be unified? `input` and `logging` both have custom tag types as well.

**Comment by timwoj** [comment]
The other thing is that `Analyzer::Tag` still gets passed down to script functions like `function name(atype: Analyzer::Tag) : string`, which throws an error if the enum value isn't an analyzer.

**Comment by timwoj** [comment]
The GTPv1 analyzer has a fairly elaborate binpac set up for parsing the protocol data, as well as a number of events in a bif file. Right now, binpac doesn’t support packet analyzers at all. In order to move the analyzer to be a packet analyzer, these set ups will need to be adapted to work with such analyzers. The following issues exist with the binpac code:

- The GTPv1_Conn class takes an Analyzer* (aliased as ZeekAnalyzer) when it is constructed. This is easily fixed by modifying src/binpac_zeek.h to add a new alias. This causes binpac to generate a class with a ZeekPacketAnalyzer in the constructor and stored inside the class.
- The GTPv1_Flow class needs to be able to retrieve the Session/Connection for the packet being processed. It uses this in order to check the encapsulation, as well as pass it into enqueue methods defined in the bif file. The encapsulation checks are easily moved back out of the pac files into the analyzer itself. The bifcl parts are harder.
- Events are generated by bifcl with signatures similar to the below. The first argument is automatically added by bifcl, but the remainder are dependent on the definition in the bif file.

		namespace zeek { namespace BifEvent {  void enqueue_gtpv1_message(zeek::analyzer::Analyzer* analyzer, zeek::Connection* c, zeek::IntrusivePtr<zeek::Val> hdr);  }  }

- The generated code only uses the GetID() method from the analyzer object. That method returns a counter value that increases every time a new session analyzer is created. This doesn’t work for packet analyzers though because packet analyzers are not uniquely generated for each session.
- All of the GTP events take a connection as their first argument. Packet analyzers generally don’t store their session, because for the most part they’re session-less. We store the related Session pointer in the packet as it’s processed, assuming it came through a transport analyzer that added one. Either way, we currently lack any way of passing that information down to the binpac code, and thus into the bifcl code. In session analyzers, there’s a Conn() method that returns the related connection. There’s nothing of the sort in packet analysis.

**Comment by timwoj** [comment]
This work was finished in https://github.com/zeek/zeek/pull/1838
